HW2/P2/parallel_vector.pyx

```diff
@@ -54,39 +54,53 @@ cpdef move_data_serial(np.int32_t[:] counts,
                        np.int32_t[:] src,
                        np.int32_t[:] dest,
                        int repeat):
-   cdef:
-       int idx, r
-
-   assert src.size == dest.size, "Sizes of src and dest arrays must match"
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+    cdef:
+        int idx, r
+ 
+    assert src.size == dest.size, "Sizes of src and dest arrays must match"
+    with nogil:
+        for r in range(repeat):
+            for idx in range(src.shape[0]):
+                if counts[src[idx]] > 0:
+                    counts[dest[idx]] += 1
+                    counts[src[idx]] -= 1
 
 
 cpdef move_data_fine_grained(np.int32_t[:] counts,
                              np.int32_t[:] src,
                              np.int32_t[:] dest,
                              int repeat):
-   cdef:
-       int idx, r
-       omp_lock_t *locks = get_N_locks(counts.shape[0])
-
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for each element of counts to parallelize
-   # data movement.  Be sure to avoid deadlock, and double-locking.
-   ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
-
-   free_N_locks(counts.shape[0], locks)
+    cdef:
+        int idx, r, lock1, lock2
+        omp_lock_t *locks = get_N_locks(counts.shape[0])
+ 
+    ##########
+    # Your code here
+    # Use parallel.prange() and a lock for each element of counts to parallelize
+    # data movement.  Be sure to avoid deadlock, and double-locking.
+    ##########
+    with nogil:
+        for r in range(repeat):
+            for idx in prange(src.shape[0], num_threads=4, schedule='dynamic'):
+                if src[idx]==dest[idx]: continue
+                if src[idx]>dest[idx]:
+                    acquire(&(locks[dest[idx]]))
+                    acquire(&(locks[src[idx]]))
+                else:
+                    acquire(&(locks[src[idx]]))
+                    acquire(&(locks[dest[idx]]))
+
+                if counts[src[idx]] > 0:
+                    counts[src[idx]] -= 1
+                    counts[dest[idx]] += 1
+                if src[idx]>dest[idx]:
+                    release(&(locks[src[idx]]))
+                    release(&(locks[dest[idx]]))
+                else:
+                    release(&(locks[dest[idx]]))
+                    release(&(locks[src[idx]]))
+                
+    free_N_locks(counts.shape[0], locks)
 
```

Comment:
You can also just lock one, add, release, and then lock the other one, substract, release. It's not really to lock both at the same time.